### ISSUE TITLE
Problems with clamped polygons on mobile devices

### ISSUE THREAD

**Issue by moisesrbd** [CesiumGS/cesium#9702]
Based on several posts that have been generated during the last few months and carrying out new tests with various mobile devices on Android, we have detected a problem with the polygons campled to ground over 3D tiles or Terrain. Polygons don't draw well when viewed on mobile devices.
When approaching (close) you can notice that areas of the polygon disappear.
More details can be seen in @tonycedeno post
[https://community.cesium.com/t/problems-with-clamped-polygons-on-mobile-devices/14627](https://community.cesium.com/t/problems-with-clamped-polygons-on-mobile-devices/14627)

Sandcastle examples:
[Original Sandcastle](https://sandcastle.cesium.com/?src=Clamp%20to%20Terrain.html)
In the following link from sandcastle you can see that when zoom-in near the polygon “Draw Polygon” option also has this erratic behavior

You can see this behavior on this Story hosted in Cesium ION:
[Story](https://cesium.com/ion/stories/viewer/?id=543fab97-f74d-42eb-9368-f772dd50340b)

**Related issues already closed:**
Some posts that could be related and that are closed: #7273 #8953 #6735

On the other hand I have noticed that using 3d-tiles with classification have a similar behavior.

**Browser:** Google Chrome
Operating System: Android 10

**Devices tested:**
- Samsung S20+
- Samsung S20 Ultra
- Huawei P20 Lite
- Samsung Galaxy Tab
 
**CesiumJS** version 1.83

Any type of help that you can offer us to solve this issue will be very helpful since the functionality of showing clamped to ground entities is one of the main ones to visualize on mobile phones.

Best
Moses

**Comment by rober72hv** [comment]
Yes, apparently it is not very easy to solve. ​But the community doesn't seem to have a solution or explanation for this yet.

A proposal, while this issue is resolved, could make a tool that allows obtaining polygons not clamped from the clamped. ​I mean if you could create a grid and get the ground points including the height (I have no idea how to do it yet, but I imagine they can be obtained). Then with those points create a substitute Polygon that is not clamped. Of course, this polygon will not have the quality of the original.
Perhaps the community may have some other ideas to help solve these use cases that, as you say, is super relevant, because visualization on mobile devices is a great need.
Important as you mention about the classification, I have tried [sandcastle ](https://sandcastle.cesium.com/?src=Classification%20Types.html) and it happens as you say, mainly when using a Textured Material.
Therefore, it not only impacts the clamped polygons, but also the classification of 3Dtiles and Terrain.

Cheers

**Comment by antoniocedeno** [comment]
Yes @moisesrbd, in fact we are facing this problem, in both, polygons and using 3d-tiles with classification. I have tried this [sandcastle](https://sandcastle.cesium.com/#c=xVfZbttGFP0VQn1xAHE0++I4QVPXXQC7Dmo3D62CYESO5GlIjkBSdhJDD/2gvvQTmh/rHS6W1aRN0wU1bFPknXP3cy91bevk2rsbVyePksrdJMeu8ZsSPeueHcwnWXd/HKrW+srV88k0ud0+eDiv5tU1YFtfuMa1++D+wj6/7IUHt/MqSTZ1cZjMJ1dtu24OZ7PWNS1aFLZpa7cOdYuyYtbbSuuQvbz2jQ/VzDaAb2bMLHNlcpVqZ1i61MSm3AqbCkoMXWCC4Tqj8B9Lhl88efbkxbcnJy++0C+OTy6+/u4M/diECjyfV53ng8+otK98uSkvstq56mJtM3dS1yEmgnTxjedqZ/PXT+tQ+sbFSFB75aqD5abKWvAxORjOPUi6OJMhn6jJXOXQuvalb/21a5DN87uzD+PR7YNOXQB99Q0ov6fTRU/uNGahakLhUBFWg2TAP+yrkL0pi6eheL0C6KPkh9v5xOfzyeF8kodsU7qqjcFPKlu67unx92enyZculK6tvWsOkzNnm03tmu7Ytatj7ruTBOH5ZDvdKVzmmDOyNKkWMks5X/DUKilSQ60lDMuMOb1v7O1PkL/kANx7+zP4Fx504nXvLZy4jTeNj2E3/W1ma+gOb6P0B6gpxsYgQoXghgmjpqng2hCFkcEaUyn1NKVcca0EEpoQIbWS04gjVCLODNVEcDrAhEFSMq61IEqMQIWE4VgTRkkP5AwMQjNxLu8MSoUIV0pJSvgA5AoxaRimRnLWIyVG1CjJiRR6RAqJBDUctDE5IIVAGg4IigVo65BaIa20oQb0yxFJEGjUgnKmRpsMwVkO7hGiOiQ4iijAwZ27MDlBAsJRwI8RKDDCWmJJtJYDjiBpsGHKSM0HIDWgixvKMGEDUFMEV8oxVoOnGkksmIFijCFScBRSygmku4fBM8QgWYIrLId6aIIYFM1Axkd7BFJIhZFCkqEchgrEtIYigOkhN1IDEFIqFIBGkwxRKCTkX8kRqZCBgBQeG0BQyJ8ymmpq2C5CbgwB2H2cgjbBoE0MSAZplkxAkcxYDCYR9AzgNVEjkCPJWJekITcQj6QM8g7SsRaxmYTCQpE7g5AIBXzBRI2uYghRx3pJTcbqcySgaaC5oC97IIbcEI2ZkEqo51tgUmlbV3tb9NyBMeHz41CEeuDS7mO9WtieUWZKcfwlgj3fxp+PJHi2z/BTu3BF8gRoPlC7Z/O/R+YuziJaGYK8Cjfwqa03DgSte9V2fqgpBqYpnZS//tJ5sgxVLyF8/SopYfg0ccp3sqtQ+zdxqxXntV/5ftydnnxxOc7A1md7ss/OLy/Pz/oI/StXnC+XMMh/FySNUaYcQxC48zq6+pnNXq7qsKnywent9vkwuHPb2ouwqTM37JfdFB93KYz2z+9OwQaw+cG9cd9tgH7f7HT1q+aPdO+Wd9RzZuuX3fL/T7eGWFgpCEkdlTzlUi5TnbM8zbkkhC0ozXO3bwzcymwOy/jLpxd/pasIp4hjAk3EFOe7IRg7Ko4Bs+MPhXEK3OOGdxVa+KJYBFvnvWZf2lXvwp+9qZRd1prhuig2sOirVedoA20TFZAPttGNz9sreChx7EfnV1ftePce2gJdgRyUiimQB2j7XkosbdHsceKrUNikhNYLv2OE/muMOD755vLk23/OCWCE+gNGdD5vP7Zdsv+rXz48h05te5iklCLYAbBxCRZYCozFfF6dhgpE0sTdAqMfK0Jhb0fJSeGuDxMaFymG9SSZphT2t/4/Bxn02t+bYndD5QNDrD/3MTPsHcSbEMrL8GcH59VkOjlq2teFe9y/TifJp76MVI5fSg4QAoqX6wL2aDNbbLKX8MqfNU3/ih1/PmlDKBYQ5+34JEkWd9k4TCJDD+DNKxn/MNI7dJKsIRJfrQ4TKN69x4tQ565Oa5v7TbMv3L5j2lfrTXvfgbGwqS38Cpqq9HleuHetpm1YQ1/tWR5Fi9C2odyXDqaPZvcTdpT768Tnj97zdTDJYDI2IFluiuLCv4GmfHw0g/PvQGPVweo5eF7Y1/HYFXl82j9ECB3N4PY+8jc) with a polygon over a tileset.

**Comment by lilleyse** [comment]
One thing to try is to switch to a WebGL 2 context.

```
var viewer = new Cesium.Viewer("cesiumContainer", {
  contextOptions: {
    requestWebgl2: true
  }
});
```

The clamping implementation works best when `gl_FragDepth` is available which is always true in the WebGL 2 context. Some mobile devices don't support the `EXT_frag_depth` extension in the WebGL 1 context.

If that doesn't help it may be a harder problem related to fragment shader precision issues for some mobile devices.

**Comment by moisesrbd** [comment]
Thanks @lilleyse I will test this option and get back to you with results.

**Comment by moisesrbd** [comment]
I did a group of tests using the option suggested by @lilleyse , the results are:
- When using Chrome the scene fails, the screen flashes white constantly. It must be that the version that i use (92.0.4515.115) does not support this context.
- Using Firefox, an improvement is noticeable, perhaps I notice a little less performance but the scene looks much better, the polygons do not disappear.

It would be interesting to know how to detect  if the browser supports this context. 
I'll keep researching these points, but if anyone in the community knows about it, it would be great.
Thanks
Moises

**Comment by moisesrbd** [comment]
New Updates.
When use 
```
var viewer = new Cesium.Viewer("cesiumContainer", {
  contextOptions: {
    requestWebgl2: true
  }
});
```
3D tiles and CZML entities that have opacity less than 1 are not shown on mobile devices with Android 10.
All entities that have transparency and 3D tiles disappear from the scene. If we don't use webgl 2, the clamped polygons disappear and the entities with opacity display correctly.
Any other ideas?

Best regards

**Comment by moisesrbd** [comment]
Tested using CesiumJS 1.84.
Same problems.

**Using Android and Firefox**
```
var viewer = new Cesium.Viewer("cesiumContainer", {
  contextOptions: {
    requestWebgl2: true
  }
});
```

- Clamped Polygons OK
- 3D Tiles and Entities with opacity < 1 disappear

**Not using  requestWebgl2: true**
- Clamped Polygons don't work
- 3D Tiles and Entities with opacity < 1  OK
